cranelift/isle/isle/src/trie_again.rs

```diff
@@ -0,0 +1,543 @@
+                // the circular linked list that points back to this binding site and make it point
+                // to `b` instead.
+                let mut cur = *entry.get();
+                loop {
+                    let next = self.equals.get_mut(&cur).unwrap();
+                    if *next == a {
+                        *next = b;
+                        break;
+                    }
+                    cur = *next;
+                }
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+struct UnmatchableError {
+    pos: Pos,
+    constraint_a: Constraint,
+    constraint_b: Constraint,
+}
+
+#[derive(Debug, Default)]
+struct RuleSetBuilder {
+    current_rule: Rule,
+    binding_map: HashMap<Binding, BindingId>,
+    expr_map: HashMap<Expr, ExprId>,
+    unmatchable: Vec<UnmatchableError>,
+    rules: RuleSet,
+}
+
+impl RuleSetBuilder {
+    fn add_rule(
+        &mut self,
+        rule: &sema::Rule,
+        termenv: &sema::TermEnv,
+        tyenv: &sema::TypeEnv,
+        errors: &mut Vec<Error>,
+    ) {
+        self.current_rule.pos = rule.pos;
+        self.current_rule.prio = rule.prio;
+        self.current_rule.result = rule.visit(self, termenv);
+        self.normalize_equivalence_classes();
+        let rule = std::mem::take(&mut self.current_rule);
+
+        if self.unmatchable.is_empty() {
+            self.rules.rules.push(rule);
+        } else {
+            // If this rule can never match, drop it so it doesn't affect overlap checking.
+            errors.extend(self.unmatchable.drain(..).map(|err| {
+                let src = Source::new(
+                    tyenv.filenames[err.pos.file].clone(),
+                    tyenv.file_texts[err.pos.file].clone(),
+                );
+                Error::UnmatchableError {
+                    msg: format!(
+                        "rule requires binding to match both {:?} and {:?}",
+                        err.constraint_a, err.constraint_b
+                    ),
+                    src,
+                    span: Span::new_single(err.pos),
+                }
+            }))
+        }
+    }
+
+    // If a binding site is constrained and also required to be equal to another binding site, then
+    // copy the constraint and push the equality inside it.
+    fn normalize_equivalence_classes(&mut self) {
+        let mut constraints = Vec::new();
+        for binding in self.current_rule.equals.keys() {
+            if let Some(constraint) = self.current_rule.constraints.get(binding) {
+                constraints.push((*binding, *constraint));
+            }
+        }
+
+        while let Some((mut binding, constraint)) = constraints.pop() {
+            struct Pending {
+                binding: BindingId,
+                variant: sema::VariantId,
+                field: TupleIndex,
+            }
+            let mut bindings = if let Constraint::Variant(fields, _, variant) = constraint {
```

Comment:
what are the "bindings" here -- can you name this variable more descriptively? They're sub-bindings (bindings whose value must be computed first), or...? Can we factor out the logic here into a helper?